### PR TITLE
Added windows support to hermetic toolchains

### PR DESCRIPTION
**Pull request by UebelAndre**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Comment by thundergolfer** [comment]
```
INFO: Repository python310_x86_64-pc-windows-msvc instantiated at:
  C:/users/user/source/repos/rulespythondev/WORKSPACE.bazel:14:27: in <toplevel>
  C:/users/user/_bazel_user/3bdyrqkj/external/rules_python/python/repositories.bzl:178:26: in python_register_toolchains
Repository rule python_repository defined at:
  C:/users/user/_bazel_user/3bdyrqkj/external/rules_python/python/repositories.bzl:134:36: in <toplevel>
WARNING: Download from https://github.com/indygreg/python-build-standalone/releases/download/20220227/cpython-3.10.2+20220227-x86_64-pc-windows-msvc-static-install_only.tar.gz failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException Checksum was a293c5838dd9c8438a84372fb95dda9752df63928a8a2ae516438f187f89567d but wanted 37e647b67a052ba576e8455d82e472ee37e8095b35387d49efe9f63a38d3f6d0
ERROR: An error occurred during the fetch of repository 'python310_x86_64-pc-windows-msvc':
   Traceback (most recent call last):
        File "C:/users/user/_bazel_user/3bdyrqkj/external/rules_python/python/repositories.bzl", line 47, column 48, in _python_repository_impl
                download_result = rctx.download_and_extract(
Error in download_and_extract: java.io.IOException: Error downloading [https://github.com/indygreg/python-build-standalone/releases/download/20220227/cpython-3.10.2+20220227-x86_64-pc-windows-msvc-static-install_only.tar.gz] to C:/users/user/_bazel_user/3bdyrqkj/external/python310_x86_64-pc-windows-msvc/temp15347681220738251273/cpython-3.10.2+20220227-x86_64-pc-windows-msvc-static-install_only.tar.gz: Checksum was a293c5838dd9c8438a84372fb95dda9752df63928a8a2ae516438f187f89567d but wanted 37e647b67a052ba576e8455d82e472ee37e8095b35387d49efe9f63a38d3f6d0
ERROR: C:/users/user/source/repos/rulespythondev/WORKSPACE.bazel:14:27: fetching python_repository rule //external:python310_x86_64-pc-windows-msvc: Traceback (most recent  call last):
        File "C:/users/user/_bazel_user/3bdyrqkj/external/rules_python/python/repositories.bzl", line 47, column 48, in _python_repository_impl
                download_result = rctx.download_and_extract(
Error in download_and_extract: java.io.IOException: Error downloading [https://github.com/indygreg/python-build-standalone/releases/download/20220227/cpython-3.10.2+20220227-x86_64-pc-windows-msvc-static-install_only.tar.gz] to C:/users/user/_bazel_user/3bdyrqkj/external/python310_x86_64-pc-windows-msvc/temp15347681220738251273/cpython-3.10.2+20220227-x86_64-pc-windows-msvc-static-install_only.tar.gz: Checksum was a293c5838dd9c8438a84372fb95dda9752df63928a8a2ae516438f187f89567d but wanted 37e647b67a052ba576e8455d82e472ee37e8095b35387d49efe9f63a38d3f6d0
ERROR: C:/users/user/source/repos/rulespythondev/BUILD.bazel:1:10: //:foo depends on @python310_x86_64-pc-windows-msvc//:python_runtimes in repository @python310_x86_64-pc-windows-msvc which failed to fetch. no such package '@python310_x86_64-pc-windows-msvc//': java.io.IOException: Error downloading [https://github.com/indygreg/python-build-standalone/releases/download/20220227/cpython-3.10.2+20220227-x86_64-pc-windows-msvc-static-install_only.tar.gz] to C:/users/user/_bazel_user/3bdyrqkj/external/python310_x86_64-pc-windows-msvc/temp15347681220738251273/cpython-3.10.2+20220227-x86_64-pc-windows-msvc-static-install_only.tar.gz: Checksum was a293c5838dd9c8438a84372fb95dda9752df63928a8a2ae516438f187f89567d but wanted 37e647b67a052ba576e8455d82e472ee37e8095b35387d49efe9f63a38d3f6d0
ERROR: Analysis of target '//:foo' failed; build aborted: Analysis failed
INFO: Elapsed time: 12.891s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (34 packages loaded, 121 targets configured)
FAILED: Build did NOT complete successfully (34 packages loaded, 121 targets configured)
```

File: python/private/versions.bzl
Comment (thundergolfer) [inline python/private/versions.bzl:24]: Can we provide these integrity strings as 'standard' hex sha256 strings? I see that the docs say use "Subresource Integrity format" but it seems like a small usability issue that the source code integrity strings are in a different format to what Bazel spits out in logs.

File: python/private/versions.bzl
Comment (thundergolfer) [inline python/private/versions.bzl:22]: @UebelAndre I think we should keep this commentary, but agree it should be `256` algorithm and not `384`. @f0rmiga why 384?

Also @f0rmiga I think we should:

* have in the commentary an explanation of the fact that SRI format is required, as that's different to other places in Bazel.
* Use --long-options in the `shasum` example command not short ones. I checked and thankfully shasum support it.
* Explain why `xxd -r -p` is needed.

**Comment by thundergolfer** [comment]
Fixing the checksum locally has the basics working 🎉 

```
$ bazel run //:foo
Loading:
Loading: 0 packages loaded
Analyzing: target //:foo (0 packages loaded, 0 targets configured)
INFO: Analyzed target //:foo (23 packages loaded, 84 targets configured).
INFO: Found 1 target...
[1 / 6] [Prepa] BazelWorkspaceStatusAction stable-status.txt
[5 / 7] PythonZipper foo.zip; 1s local
Target //:foo up-to-date:
  bazel-bin/foo.exe
  bazel-bin/foo.zip
INFO: Elapsed time: 3.301s, Critical Path: 2.74s
INFO: 2 processes: 1 internal, 1 local.
INFO: Build completed successfully, 2 total actions
INFO: Running command line: bazel-bin/foo.exe
INFO: Build completed successfully, 2 total actions
sys.version='3.10.2 (main, Feb 27 2022, 18:43:34) [MSC v.1929 64 bit (AMD64)]'
sys.executable='C:\\Users\\User\\AppData\\Local\\Temp\\Bazel.runfiles_pj0rlg5e\\runfiles\\python310_x86_64-pc-windows-msvc\\python.exe'
hello world
```

**Comment by UebelAndre** [comment]
@thundergolfer  in case it's not obvious, this PR is targeting https://github.com/bazelbuild/rules_python/pull/618

**Comment by UebelAndre** [comment]
> I'm hitting the following minor problem when I run the PR:
> 
> ```
> INFO: Repository python310_x86_64-pc-windows-msvc instantiated at:
>   C:/users/user/source/repos/rulespythondev/WORKSPACE.bazel:14:27: in <toplevel>
>   C:/users/user/_bazel_user/3bdyrqkj/external/rules_python/python/repositories.bzl:178:26: in python_register_toolchains
> Repository rule python_repository defined at:
>   C:/users/user/_bazel_user/3bdyrqkj/external/rules_python/python/repositories.bzl:134:36: in <toplevel>
> WARNING: Download from https://github.com/indygreg/python-build-standalone/releases/download/20220227/cpython-3.10.2+20220227-x86_64-pc-windows-msvc-static-install_only.tar.gz failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException Checksum was a293c5838dd9c8438a84372fb95dda9752df63928a8a2ae516438f187f89567d but wanted 37e647b67a052ba576e8455d82e472ee37e8095b35387d49efe9f63a38d3f6d0
> ERROR: An error occurred during the fetch of repository 'python310_x86_64-pc-windows-msvc':
>    Traceback (most recent call last):
>         File "C:/users/user/_bazel_user/3bdyrqkj/external/rules_python/python/repositories.bzl", line 47, column 48, in _python_repository_impl
>                 download_result = rctx.download_and_extract(
> Error in download_and_extract: java.io.IOException: Error downloading [https://github.com/indygreg/python-build-standalone/releases/download/20220227/cpython-3.10.2+20220227-x86_64-pc-windows-msvc-static-install_only.tar.gz] to C:/users/user/_bazel_user/3bdyrqkj/external/python310_x86_64-pc-windows-msvc/temp15347681220738251273/cpython-3.10.2+20220227-x86_64-pc-windows-msvc-static-install_only.tar.gz: Checksum was a293c5838dd9c8438a84372fb95dda9752df63928a8a2ae516438f187f89567d but wanted 37e647b67a052ba576e8455d82e472ee37e8095b35387d49efe9f63a38d3f6d0
> ERROR: C:/users/user/source/repos/rulespythondev/WORKSPACE.bazel:14:27: fetching python_repository rule //external:python310_x86_64-pc-windows-msvc: Traceback (most recent  call last):
>         File "C:/users/user/_bazel_user/3bdyrqkj/external/rules_python/python/repositories.bzl", line 47, column 48, in _python_repository_impl
>                 download_result = rctx.download_and_extract(
> Error in download_and_extract: java.io.IOException: Error downloading [https://github.com/indygreg/python-build-standalone/releases/download/20220227/cpython-3.10.2+20220227-x86_64-pc-windows-msvc-static-install_only.tar.gz] to C:/users/user/_bazel_user/3bdyrqkj/external/python310_x86_64-pc-windows-msvc/temp15347681220738251273/cpython-3.10.2+20220227-x86_64-pc-windows-msvc-static-install_only.tar.gz: Checksum was a293c5838dd9c8438a84372fb95dda9752df63928a8a2ae516438f187f89567d but wanted 37e647b67a052ba576e8455d82e472ee37e8095b35387d49efe9f63a38d3f6d0
> ERROR: C:/users/user/source/repos/rulespythondev/BUILD.bazel:1:10: //:foo depends on @python310_x86_64-pc-windows-msvc//:python_runtimes in repository @python310_x86_64-pc-windows-msvc which failed to fetch. no such package '@python310_x86_64-pc-windows-msvc//': java.io.IOException: Error downloading [https://github.com/indygreg/python-build-standalone/releases/download/20220227/cpython-3.10.2+20220227-x86_64-pc-windows-msvc-static-install_only.tar.gz] to C:/users/user/_bazel_user/3bdyrqkj/external/python310_x86_64-pc-windows-msvc/temp15347681220738251273/cpython-3.10.2+20220227-x86_64-pc-windows-msvc-static-install_only.tar.gz: Checksum was a293c5838dd9c8438a84372fb95dda9752df63928a8a2ae516438f187f89567d but wanted 37e647b67a052ba576e8455d82e472ee37e8095b35387d49efe9f63a38d3f6d0
> ERROR: Analysis of target '//:foo' failed; build aborted: Analysis failed
> INFO: Elapsed time: 12.891s
> INFO: 0 processes.
> FAILED: Build did NOT complete successfully (34 packages loaded, 121 targets configured)
> FAILED: Build did NOT complete successfully (34 packages loaded, 121 targets configured)
> ```
> 
> Looks like a straightforward hash mismatch.

@thundergolfer perhaps you're running into https://github.com/indygreg/python-build-standalone/issues/116#issuecomment-1053640069

File: python/private/versions.bzl
Comment (UebelAndre) [reply python/private/versions.bzl:24]: I see no reason why these couldn't become `sha256` values. It would just require updating the attribute in `repositories.bzl` I leave this decision up to @f0rmiga https://github.com/bazelbuild/rules_python/pull/618#discussion_r815470830

File: python/private/versions.bzl
Comment (UebelAndre) [reply python/private/versions.bzl:22]: I have no attachment as to whether or not this stays or is removed. I think the repository rule should report what value this should be should one not be provided though (which is something I fixed in this PR). Since this PR targets https://github.com/bazelbuild/rules_python/pull/618 I figured it'd be up to @f0rmiga as to what gets accepted and what doesn't.

**Comment by UebelAndre** [comment]
> Looks like a straightforward hash mismatch.

Updated the checksums

File: python/private/versions.bzl
Comment (f0rmiga) [reply python/private/versions.bzl:22]: I'm okay with dropping the comment. Anyone dealing with Bazel at this level should know how to grab the hashes (there are many ways to accomplish this).

I'm also not particularly attached to which algorithm to use - this is an integrity check, not a user password in a database.

File: python/repositories.bzl
Comment (f0rmiga) [inline python/repositories.bzl:36]: I'm not a Windows user. Would a static binary for Windows have the same limitation as a fully static binary for Linux that a shared object (i.e. a native Python extension) cannot be loaded at runtime (not able to do dlopen)?

File: python/repositories.bzl
Comment (f0rmiga) [inline python/repositories.bzl:95]: A `select` with the platforms would be nicer here.

File: python/repositories.bzl
Comment (f0rmiga) [inline python/repositories.bzl:136]:
```suggestion
    doc = "Fetches the external tools needed for the Python toolchain.",
```

File: python/repositories.bzl
Comment (f0rmiga) [inline python/repositories.bzl:139]:
```suggestion
            doc = "The integrity hash for the Python interpreter tarball.",
```

File: python/repositories.bzl
Comment (f0rmiga) [inline python/repositories.bzl:143]:
```suggestion
            doc = "The platform name for the Python interpreter tarball.",
```

File: python/repositories.bzl
Comment (f0rmiga) [inline python/repositories.bzl:148]:
```suggestion
            doc = "The Python version.",
```

**Review by f0rmiga — CHANGES_REQUESTED** [review]
Awesome! Thanks for working on the Windows portion.

File: python/repositories.bzl
Comment (UebelAndre) [reply python/repositories.bzl:36]: I'm not sure myself, but this was the only artifact that worked. My thought was to hopefully see something merged that works in some cases and fix other issues as they're reported.

Comment (f0rmiga) [reply python/repositories.bzl:36]: Well, this could possibly be a blocker for merging my PR into main. I'll take this for now, but will have to investigate before releasing this feature.

**Review by f0rmiga — APPROVED** [review]

File: python/private/versions.bzl
Comment (thundergolfer) [reply python/private/versions.bzl:22]: @f0rmiga I think we _should keep the comment_. As a Bazel maintainer I can figure out how to generate the integrity strings, but it's just easier if the code documents this as it's not the trivial `shasum -a 256 foo` uses in other places in Bazel. 

I liked the commentary, and wanted it expanded address the bullet points above.

Comment (f0rmiga) [reply python/private/versions.bzl:22]: https://github.com/bazelbuild/rules_python/pull/618/commits/d7ac042f2ad92670f60e25d4f97fcdef567c7859 addresses this more elegantly.

Comment (thundergolfer) [reply python/private/versions.bzl:22]: Yep, looks good.

**Pull request merged**
